### PR TITLE
Merge master 1.4.1 into ergon master

### PR DESCRIPTION
**Pull request by shanselm-ergon**

### CODE REVIEW

File: android/src/main/java/com/getcapacitor/plugin/http/FilesystemUtils.java [hunk 1]
```diff
@@ -16,7 +16,7 @@ public class FilesystemUtils {
     public static final String DIRECTORY_EXTERNAL_STORAGE = "EXTERNAL_STORAGE";
 
     public static File getFileObject(Context c, String path, String directory) {
-        if (directory == null) {
+        if (directory == null || path.startsWith("file://")) {
```
Comment (shanselm-ergon) [inline android/src/main/java/com/getcapacitor/plugin/http/FilesystemUtils.java:19]: not sure if it should be an `&&` instead? Could be merged.

File: android/src/main/java/com/getcapacitor/plugin/http/Http.java [hunk 2]
```diff
@@ -260,23 +268,18 @@ public void deleteCookie(PluginCall call) {
...
-        Log.i(getLogTag(), "Clear cookies for url \"urlString\"");
-        // TODO
-        // This is only checks if the string starts with http:// or https://.
-        // Maybe has to be changed when known how the method for removing cookies by given URL is implemented.
-        if (isValidURL(urlString)) {
-            // TODO: add support for clearing cookies for a given URL
-            call.reject("clearCookies for a specific URL is not yet supported");
-        } else {
-            call.reject("Could not parse URL. Check that \"url\" has valid format");
+        String url = getServerUrl(call);
```
Comment (shanselm-ergon) [inline android/src/main/java/com/getcapacitor/plugin/http/Http.java:271]: Looks ok

File: android/src/main/java/com/getcapacitor/plugin/http/HttpRequestHandler.java [hunk 3]
```diff
@@ -126,6 +126,11 @@ public HttpURLConnectionBuilder setUrlParams(JSObject params) throws URISyntaxEx
             String initialQueryBuilderStr = initialQuery == null ? "" : initialQuery;
 
             Iterator<String> keys = params.keys();
+
```
Comment (shanselm-ergon) [inline android/src/main/java/com/getcapacitor/plugin/http/HttpRequestHandler.java:129]: Looks ok

File: android/src/main/java/com/getcapacitor/plugin/http/HttpRequestHandler.java [hunk 4]
```diff
@@ -157,8 +162,19 @@ public HttpURLConnectionBuilder setUrlParams(JSObject params) throws URISyntaxEx
             String urlQuery = urlQueryBuilder.toString();
 
             URI uri = url.toURI();
-            URI encodedUri = new URI(uri.getScheme(), uri.getAuthority(), uri.getPath(), urlQuery, uri.getFragment());
-            this.url = encodedUri.toURL();
+            if (shouldEncode) {
```
Comment (shanselm-ergon) [inline android/src/main/java/com/getcapacitor/plugin/http/HttpRequestHandler.java:165]: Looks ok

File: ios/Plugin/CapacitorCookieManager.swift [hunk 5]
```diff
@@ -46,4 +46,9 @@ public class CapacitorCookieManager {
         jar.cookies?.forEach { jar.deleteCookie($0) }
         os_log("%{public}@", log: HttpPlugin.logTarget, type: OSLogType.info, "All cookies successfully deleted")
     }
+    
+    public func clearAllCookies() {
```
Comment (shanselm-ergon) [inline ios/Plugin/CapacitorCookieManager.swift:50]: check with our implementation

File: ios/Plugin/CapacitorUrlRequest.swift [hunk 6]
```diff
@@ -13,15 +13,23 @@ public class CapacitorUrlRequest: NSObject, URLSessionTaskDelegate {
...
+                headers["Accept-Language"] = "\(lang);q=0.5"
+            }
+            request.addValue(headers["Accept-Language"]!, forHTTPHeaderField: "Accept-Language")
+        }
     }
     
     private func getRequestDataAsJson(_ data: JSValue) throws -> Data? {
-        guard let convertedData = JSONDataConverter.convert(data) else {
-            throw CapacitorUrlRequestError.serializationError("[ data ] argument for request with content-type [ application/json ] contains invalid values")
+        // We need to check if the JSON is valid before attempting to serialize, as JSONSerialization.data will not throw an exception that can be caught, and will cause the application to crash if it fails.
```
Comment (shanselm-ergon) [inline ios/Plugin/CapacitorUrlRequest.swift:27]: looks ok

File: ios/Plugin/CapacitorUrlRequest.swift [hunk 7]
```diff
@@ -88,24 +96,26 @@ public class CapacitorUrlRequest: NSObject, URLSessionTaskDelegate {
     }
     
     func getRequestData(_ body: JSValue, _ contentType: String) throws -> Data? {
-        if contentType.contains("application/json") {
+        // If data can be parsed directly as a string, return that without processing.
+        if let strVal = try? getRequestDataAsString(body) {
```
Comment (shanselm-ergon) [inline ios/Plugin/CapacitorUrlRequest.swift:100]: Not sure if this is correct, or better said, what `getRequestDataAsString` does. JSON is for me also a valid string value and would be treated as string here.

File: ios/Plugin/CapacitorUrlRequest.swift [hunk 8]
```diff
@@ -88,24 +96,26 @@ public class CapacitorUrlRequest: NSObject, URLSessionTaskDelegate {
...
-            return try getRequestDataAsString(body)
+            throw CapacitorUrlRequestError.serializationError("[ data ] argument could not be parsed for content type [ \(contentType) ]")
         }
     }
 
     public func setRequestHeaders(_ headers: [String: String]) {
         headers.keys.forEach { (key: String) in
             let value = headers[key]
             request.addValue(value!, forHTTPHeaderField: key)
+            self.headers[key] = value
```
Comment (shanselm-ergon) [inline ios/Plugin/CapacitorUrlRequest.swift:117]: ok

File: ios/Plugin/Plugin.swift [hunk 9]
```diff
@@ -127,39 +139,26 @@ import os.log
...
-        
+
         call.resolve()
     }
 
     @objc func clearCookies(_ call: CAPPluginCall) {
-        guard let url = call.getServerURLOrReject() else { return }
-        cookieManager!.clearCookies(for: url)
-        call.resolve()
+        let url = getServerUrl(call)
```
Comment (shanselm-ergon) [inline ios/Plugin/Plugin.swift:152]: looks ok

File: ios/Plugin/CapacitorCookieManager.swift (hunk 5 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorCookieManager.swift:50]: skip

File: ios/Plugin/CapacitorUrlRequest.swift [hunk 10]
```diff
@@ -13,15 +13,23 @@ public class CapacitorUrlRequest: NSObject, URLSessionTaskDelegate {
         request = URLRequest(url: url)
         request.httpMethod = method
         headers = [:]
+        if let lang = Locale.autoupdatingCurrent.languageCode {
```
Comment (7-plus-t) [inline ios/Plugin/CapacitorUrlRequest.swift:16]: probably ok

File: ios/Plugin/CapacitorUrlRequest.swift [hunk 11]
```diff
@@ -13,15 +13,23 @@ public class CapacitorUrlRequest: NSObject, URLSessionTaskDelegate {
...
+                headers["Accept-Language"] = "\(lang)-\(country),\(lang);q=0.5"
+            } else {
+                headers["Accept-Language"] = "\(lang);q=0.5"
+            }
+            request.addValue(headers["Accept-Language"]!, forHTTPHeaderField: "Accept-Language")
+        }
     }
     
     private func getRequestDataAsJson(_ data: JSValue) throws -> Data? {
-        guard let convertedData = JSONDataConverter.convert(data) else {
```
Comment (7-plus-t) [inline ios/Plugin/CapacitorUrlRequest.swift:19]: How come the conversion is no longer needed?

File: ios/Plugin/CapacitorUrlRequest.swift (hunk 8 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorUrlRequest.swift:117]: it changes the behaviour from replacing the whole array on the instance to  only changing the ones passed as param

File: ios/Plugin/HttpRequestHandler.swift [hunk 12]
```diff
@@ -177,7 +176,7 @@ class HttpRequestHandler {
         let timeout = (connectTimeout ?? readTimeout ?? 600000.0) / 1000.0
         request.setTimeout(timeout)
 
-        if let data = call.jsObjectRepresentation["data"] {
+        if let data = call.options["data"] as? JSValue {
```
Comment (7-plus-t) [inline ios/Plugin/HttpRequestHandler.swift:179]: (options as? [String: JSValue] ?? [:])["data"]
vs.
options["data"] as? JSValue

the old version seems more robust?

File: ios/Plugin/HttpRequestHandler.swift [hunk 13]
```diff
@@ -192,7 +191,6 @@ class HttpRequestHandler {
             urlSession.finishTasksAndInvalidate()
 
             if error != nil {
-                call.reject("Error", "REQUEST", error, [:])
```
Comment (7-plus-t) [inline ios/Plugin/HttpRequestHandler.swift:195]: not ok

File: ios/Plugin/Plugin.swift [hunk 14]
```diff
@@ -98,9 +98,21 @@ import os.log
         call.resolve()
     }
 
+    @objc func getCookiesMap(_ call: CAPPluginCall) {
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:101]: needs improvements. not ok as is

File: ios/Plugin/Plugin.swift (hunk 9 above)
Comment (7-plus-t) [reply ios/Plugin/Plugin.swift:152]: our version is more robust, also this misses a resolve?

File: ios/Plugin/Plugin.swift [hunk 15]
```diff
@@ -127,39 +139,26 @@ import os.log
...
-    
+
     @objc func clearAllCookies(_ call: CAPPluginCall) {
         os_log("%{public}@", log: HttpPlugin.logTarget, type: OSLogType.info, "Clear all cookies")
         cookieManager!.clearAllCookies()
         call.resolve()
     }
 }
-
-private extension CAPPluginCall {
```
Comment (7-plus-t) [inline ios/Plugin/Plugin.swift:152]: keep

File: ios/Plugin/CapacitorUrlRequest.swift (hunk 11 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorUrlRequest.swift:19]: skip

File: ios/Plugin/CapacitorUrlRequest.swift (hunk 10 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorUrlRequest.swift:16]: skip

File: ios/Plugin/CapacitorUrlRequest.swift (hunk 7 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorUrlRequest.swift:100]: skip

File: ios/Plugin/CapacitorUrlRequest.swift (hunk 8 above)
Comment (7-plus-t) [reply ios/Plugin/CapacitorUrlRequest.swift:117]: skip

File: ios/Plugin/HttpRequestHandler.swift (hunk 12 above)
Comment (7-plus-t) [reply ios/Plugin/HttpRequestHandler.swift:179]: skip

File: ios/Plugin/HttpRequestHandler.swift (hunk 13 above)
Comment (7-plus-t) [reply ios/Plugin/HttpRequestHandler.swift:195]: skip

File: ios/Plugin/Plugin.swift (hunk 14 above)
Comment (7-plus-t) [reply ios/Plugin/Plugin.swift:101]: skip

File: ios/Plugin/Plugin.swift (hunk 9 above)
Comment (7-plus-t) [reply ios/Plugin/Plugin.swift:152]: skip

**Comment by shanselm-ergon** [comment]
We don't merge these changes, there is no benefit for our fork

**Pull request closed without merging**